src/main/kotlin/org/elm/workspace/compiler/ElmBuildAction.kt

```diff
@@ -0,0 +1,95 @@
+import org.elm.workspace.ElmProject
+import org.elm.workspace.elmToolchain
+import org.elm.workspace.elmWorkspace
+
+
+class ElmBuildAction : AnAction() {
+
+    private val elmMainTypes: Set<Pair<String, String>> = hashSetOf(Pair("Platform", "Program"), Pair("Html", "Html"))
+
+    private val elmJsonReport = ElmJsonReport()
+
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.project
+                ?: return
+        saveAllDocuments()
+
+        val elmCLI = project.elmToolchain.elmCLI
+        val fixAction = "Fix" to { project.elmWorkspace.showConfigureToolchainUI() }
+        if (elmCLI == null) {
+            project.showBalloon("Could not find elm", NotificationType.ERROR, fixAction)
+            return
+        }
+
+        val elmProject = project.elmWorkspace.allProjects.first()
+
+        // find "main" function
+        val clientLocation = LookupClientLocation(project, elmProject)
+        val elements = ElmLookup.findByName<ElmNamedElement>("main", clientLocation)
+        val mainElements = elements
+                .filter { element -> element.findInference() != null && element.findInference()?.ty is TyUnion }
+                .filter { element -> elmMainTypes.contains(Pair((element.findInference()?.ty as TyUnion).module, (element.findInference()?.ty as TyUnion).name)) }
+
+        if (mainElements.isNotEmpty()) {
+            val path = mainElements[0].containingFile.virtualFile.path
+            try {
+                val json = elmCLI.make(project, elmProject, path).stderr
+                if (json.isNotEmpty()) {
+                    // TODO test _list_ of errors (only produced, if multiple independent erroneous modules are compiled.. examples for this ?)
+                    val messages = elmJsonReport.elmToCompilerMessages(json).sortedWith(compareBy({it.name}, {it.messageWithRegion.region.start.line}, {it.messageWithRegion.region.start.column}))
+                    project.messageBus.syncPublisher(ERRORS_TOPIC).update(messages)
+
+                } else {
+                    project.messageBus.syncPublisher(ERRORS_TOPIC).update(emptyList())
+                }
+                // show toolwindow
+                ToolWindowManager.getInstance(project).getToolWindow("Elm Compiler").show(null)
+            } catch (e: ExecutionException) {
+                project.showBalloon("Invalid path for 'elm' executable", NotificationType.ERROR, fixAction)
+                return
+            }
+        } else {
+            showDialog(project)
+        }
+    }
+
+    private fun showDialog(project: Project) {
+        val statusBar = WindowManager.getInstance().getStatusBar(project)
+        JBPopupFactory.getInstance().createHtmlTextBalloonBuilder("No Type Signature found. Please specify one, so that Elm Build works.", MessageType.ERROR, null).setFadeoutTime(5000).createBalloon().show(RelativePoint.getNorthEastOf(statusBar.component), Balloon.Position.atRight)
```

Comment:
Change to: `Cannot find your Elm app's main entry point. Please make sure that it has a type annotation.`